diff --ruN a/stablehlo/stablehlo/transforms/StablehloAggressiveFolder.cpp b/stablehlo/stablehlo/transforms/StablehloAggressiveFolder.cpp
--- stablehlo/stablehlo/transforms/StablehloAggressiveFolder.cpp
+++ stablehlo/stablehlo/transforms/StablehloAggressiveFolder.cpp
@@ -79,8 +79,9 @@
   } else {
     llvm::report_fatal_error("expected integer type");
   }
-  return APSInt({/*numBits=*/numBits, value},
-                /*isUnsigned=*/isUnsigned);
+  return APSInt(
+      {/*numBits=*/numBits, value, /*isSigned=*/false, /*implicitTrunc=*/true},
+      /*isUnsigned=*/isUnsigned);
 }
 
 LogicalResult validateResultTypeForEval(PatternRewriter& rewriter,

